File: frontend/src/stores/calls.ts
/**
 * P3: Calls & SMS Store
 */
import { create } from 'zustand';
import { api } from '@/lib/api';
import { getApiErrorMessage } from '@/lib/api-errors';

// Types
export interface CallLog {
  id: string;
  locationId: string;
  callerPhone: string;
  maskedPhone: string;
  direction: 'inbound' | 'outbound';
  status: 'answered' | 'missed' | 'busy' | 'failed';
  durationSeconds?: number;
  callStartedAt: string;
  callEndedAt?: string;
  textBackSent: boolean;
  textBackAt?: string;
  createdAt: string;
}

export interface SMSThread {
  id: string;
  locationId: string;
  customerPhone: string;
  maskedPhone: string;
  customerName?: string;
  status: 'active' | 'archived' | 'blocked';
  unreadCount: number;
  lastMessageAt: string;
  lastMessagePreview: string;
}

export interface SMSMessage {
  id: string;
  direction: 'inbound' | 'outbound';
  body: string;
  sentAt: string;
  status: 'queued' | 'sent' | 'delivered' | 'failed' | 'received';
  isRead: boolean;
}

export interface TextBackSettings {
  id: string;
  locationId: string;
  enabled: boolean;
  delaySeconds: number;
  respectBusinessHours: boolean;
  defaultMessage: string;
  afterHoursMessage: string;
  enableQuickReplies: boolean;
  quickReplyOptions: string[];
  twilioNumber?: string;
  forwardingNumber?: string;
}

export interface CallStats {
  totalCalls: number;
  answeredCalls: number;
  missedCalls: number;
  textBacksSent: number;
  textBackRate: number;
}

interface CallsState {
  // Data
  callLogs: CallLog[];
  threads: SMSThread[];
  messages: SMSMessage[];
  settings: TextBackSettings | null;
  stats: CallStats | null;

  // Selected
  selectedThread: SMSThread | null;

  // Loading
  isLoading: boolean;
  isLoadingMessages: boolean;
  isSending: boolean;

  // Counts
  totalUnread: number;

  // Error
  error: string | null;

  // Actions
  fetchCallLogs: (locationId: string, days?: number) => Promise<void>;
  fetchStats: (locationId: string, days?: number) => Promise<void>;

  fetchThreads: (locationId: string, unreadOnly?: boolean) => Promise<void>;
  fetchMessages: (locationId: string, threadId: string) => Promise<void>;
  sendMessage: (locationId: string, threadId: string, body: string) => Promise<void>;
  markThreadRead: (locationId: string, threadId: string) => Promise<void>;

  fetchSettings: (locationId: string) => Promise<void>;
  updateSettings: (locationId: string, data: Partial<TextBackSettings>) => Promise<void>;

  selectThread: (thread: SMSThread | null) => void;
  clearError: () => void;
}

export const useCallsStore = create<CallsState>((set, get) => ({
  // Initial state
  callLogs: [],
  threads: [],
  messages: [],
  settings: null,
  stats: null,
  selectedThread: null,
  isLoading: false,
  isLoadingMessages: false,
  isSending: false,
  totalUnread: 0,
  error: null,

  // Call Logs
  fetchCallLogs: async (locationId: string, days = 30) => {
    set({ isLoading: true, error: null });
    try {
      const response = await api.get(`/calls/${locationId}/logs`, {
        params: { days },
      });
      set({
        callLogs: response.data.items,
        isLoading: false
      });
    } catch (error) {
      set({
        error: getApiErrorMessage(error, 'Failed to fetch call logs'),
        isLoading: false
      });
    }
  },

  fetchStats: async (locationId: string, days = 30) => {
    try {
      const response = await api.get(`/calls/${locationId}/stats`, {
        params: { days },
      });
      set({ stats: response.data });
    } catch (error) {
      set({ error: getApiErrorMessage(error, 'Failed to fetch call stats') });
    }
  },

  // SMS Threads
  fetchThreads: async (locationId: string, unreadOnly = false) => {
    set({ isLoading: true, error: null });
    try {
      const response = await api.get(`/calls/${locationId}/threads`, {
        params: { unread_only: unreadOnly },
      });
      set({
        threads: response.data.items,
        totalUnread: response.data.totalUnread,
        isLoading: false
      });
    } catch (error) {
      set({
        error: getApiErrorMessage(error, 'Failed to fetch threads'),
        isLoading: false
      });
    }
  },

  fetchMessages: async (locationId: string, threadId: string) => {
    set({ isLoadingMessages: true });
    try {
      const response = await api.get(`/calls/${locationId}/threads/${threadId}`);
      set({
        messages: response.data.messages,
        isLoadingMessages: false
      });
      // Mark as read
      await get().markThreadRead(locationId, threadId);
    } catch (error) {
      set({
        error: getApiErrorMessage(error, 'Failed to fetch messages'),
        isLoadingMessages: false
      });
    }
  },

  sendMessage: async (locationId: string, threadId: string, body: string) => {
    set({ isSending: true, error: null });
    try {
      const response = await api.post(`/calls/${locationId}/threads/${threadId}/send`, {
        body,
      });
      // Add message to list
      set((state) => ({
        messages: [...state.messages, {
          id: response.data.messageId,
          direction: 'outbound',
          body,
          sentAt: new Date().toISOString(),
          status: response.data.status,
          isRead: true,
        }],
        isSending: false,
      }));
    } catch (error) {
      set({
        error: getApiErrorMessage(error, 'Failed to send message'),
        isSending: false
      });
      throw error;
    }
  },

  markThreadRead: async (locationId: string, threadId: string) => {
    try {
      await api.post(`/calls/${locationId}/threads/${threadId}/read`);
      set((state) => ({
        threads: state.threads.map((t) =>
          t.id === threadId ? { ...t, unreadCount: 0 } : t
        ),
        totalUnread: Math.max(0, state.totalUnread - (
          state.threads.find((t) => t.id === threadId)?.unreadCount || 0
        )),
      }));
    } catch {
      // Silent fail for mark read
    }
  },

  // Settings
  fetchSettings: async (locationId: string) => {
    try {
      const response = await api.get(`/calls/${locationId}/settings`);
      set({ settings: response.data });
    } catch (error) {
      set({ error: getApiErrorMessage(error, 'Failed to fetch call settings') });
    }
  },

  updateSettings: async (locationId: string, data: Partial<TextBackSettings>) => {
    try {
      const response = await api.put(`/calls/${locationId}/settings`, data);
      set({ settings: response.data });
    } catch (error) {
      set({ error: getApiErrorMessage(error, 'Failed to update call settings') });
      throw error;
    }
  },

  selectThread: (thread) => set({ selectedThread: thread, messages: [] }),
  clearError: () => set({ error: null }),
}));

File: app/schemas/calls.py
"""P3: Call & SMS schemas."""

from datetime import datetime
from typing import Optional, Literal
from uuid import UUID

from pydantic import BaseModel, Field


# ====================
# Call Log Schemas
# ====================

class CallLogResponse(BaseModel):
    """Call log entry."""
    id: UUID
    location_id: UUID
    caller_phone: str
    masked_phone: str  # For privacy display
    direction: str  # inbound, outbound
    status: str  # missed, answered, voicemail
    duration_seconds: Optional[int]
    call_started_at: datetime
    call_ended_at: Optional[datetime]
    text_back_sent: bool
    text_back_at: Optional[datetime]
    text_back_response: Optional[str]
    recording_url: Optional[str]
    notes: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


class CallLogList(BaseModel):
    """List of call logs."""
    items: list[CallLogResponse]
    total: int
    total_missed: int
    total_answered: int
    text_back_rate: float  # % of missed calls with text back


# ====================
# Text Back Settings
# ====================

class TextBackSettingsCreate(BaseModel):
    """Create text back settings for location."""
    enabled: bool = True
    
    # Response time
    delay_seconds: int = Field(
        default=60,
        ge=30,
        le=300,
        description="Seconds to wait before sending (30-300)"
    )
    
    # Business hours only
    respect_business_hours: bool = True
    
    # Templates
    default_message: str = Field(
        default="안녕하세요! 방금 전화 주셨군요. 지금 통화가 어려워 문자 드립니다. "
                "어떤 용건이신가요? 빠르게 답변 드리겠습니다.",
        max_length=500,
    )
    
    after_hours_message: Optional[str] = Field(
        default="안녕하세요! 현재 영업시간 외입니다. "
                "내일 영업 시작하면 바로 연락 드리겠습니다. "
                "급한 용건은 문자로 남겨주세요!",
        max_length=500,
    )
    
    # Quick replies
    enable_quick_replies: bool = True
    quick_reply_options: list[str] = Field(
        default=[
            "예약 문의",
            "가격 문의",
            "영업시간 문의",
            "기타 문의",
        ]
    )


class TextBackSettingsUpdate(BaseModel):
    """Update text back settings."""
    enabled: Optional[bool] = None
    delay_seconds: Optional[int] = Field(None, ge=30, le=300)
    respect_business_hours: Optional[bool] = None
    default_message: Optional[str] = Field(None, max_length=500)
    after_hours_message: Optional[str] = Field(None, max_length=500)
    enable_quick_replies: Optional[bool] = None
    quick_reply_options: Optional[list[str]] = None


class TextBackSettingsResponse(BaseModel):
    """Text back settings response."""
    id: UUID
    location_id: UUID
    enabled: bool
    delay_seconds: int
    respect_business_hours: bool
    default_message: str
    after_hours_message: Optional[str]
    enable_quick_replies: bool
    quick_reply_options: list[str]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ====================
# SMS Thread Schemas
# ====================

class SMSMessage(BaseModel):
    """Single SMS message."""
    id: UUID
    direction: str  # inbound, outbound
    body: str
    sent_at: datetime
    status: str  # sent, delivered, failed
    is_read: bool


class SMSThreadResponse(BaseModel):
    """SMS conversation thread."""
    id: UUID
    location_id: UUID
    customer_phone: str
    masked_phone: str
    customer_name: Optional[str]
    last_message_at: datetime
    last_message_preview: str
    unread_count: int
    messages: list[SMSMessage] = []

    class Config:
        from_attributes = True


class SMSThreadList(BaseModel):
    """List of SMS threads."""
    items: list[SMSThreadResponse]
    total: int
    total_unread: int


class SMSSendRequest(BaseModel):
    """Send SMS in thread."""
    body: str = Field(..., min_length=1, max_length=1600)


class SMSSendResponse(BaseModel):
    """Send SMS response."""
    message_id: UUID
    status: str
    sent_at: datetime


# ====================
# Analytics Schemas
# ====================

class CallAnalytics(BaseModel):
    """Call analytics summary."""
    period_start: datetime
    period_end: datetime
    
    # Call stats
    total_calls: int
    total_missed: int
    total_answered: int
    miss_rate: float
    
    # Text back stats
    text_backs_sent: int
    text_back_rate: float
    response_rate: float  # Customer replies to text back
    
    # Conversion
    estimated_saves: int  # Customers who might have been lost
    save_rate: float
    
    # Time breakdown
    busiest_hour: int
    busiest_day: str
    calls_by_hour: dict[int, int]
    calls_by_day: dict[str, int]


class CallStatsDaily(BaseModel):
    """Daily call stats for charts."""
    date: str
    total: int
    missed: int
    answered: int
    text_backs: int
